### PR TITLE
[interpreter] more updates around delegates, generics, floats in icall trampoline, ...

### PR DESCRIPTION
**Pull request by lewurm**
noteable changes:
* reuse managed wrapper from bitcode for `delegate.invoke ()` and existing icall for delegate ctor.
* implement callback `mono_walk_stack_with_ctx`.
* added more regression tests, with this PR we're from 441 to 867 passing tests.

rest are minor fixes for corner cases.

@vargaz please review changes in `metadata/marshal.c`.

current state:
```
MONO_PATH=/Users/bernhardu/work/mono/mcs/class/lib/net_4_x ../../runtime/mono-wrapper --interpreter --regression basic.exe basic-float.exe basic-long.exe basic-calls.exe objects.exe arrays.exe basic-math.exe exceptions.exe devirtualization.exe generics.exe basic-simd.exe basic-vectors.exe
Test run: image=/Users/bernhardu/work/mono/mono/mini/basic.exe
Results: total tests: 134, all pass
Elapsed time: 0.001596 secs (0.001596, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/basic-float.exe
Results: total tests: 54, all pass
Elapsed time: 0.000480 secs (0.000480, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/basic-long.exe
Results: total tests: 97, all pass
Elapsed time: 0.000689 secs (0.000689, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/basic-calls.exe
Results: total tests: 27, all pass
Elapsed time: 0.000407 secs (0.000407, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/objects.exe
Results: total tests: 90, all pass
Elapsed time: 0.091792 secs (0.091792, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/arrays.exe
skip test_0_multi_dimension_arrays...
skip test_0_multidym_array_with_negative_lower_bound...
skip test_0_newarr_ovf...
Results: total tests: 30, all pass
Elapsed time: 0.017633 secs (0.017633, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/basic-math.exe
Results: total tests: 20, all pass
Elapsed time: 0.001138 secs (0.001138, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/exceptions.exe
skip test_0_int_cast...
skip test_0_div_zero...
skip test_0_long_div_zero...
skip test_3_checked_cast_un...
skip test_4_checked_cast...
skip test_0_multi_dim_array_access...
skip test_2_array_mismatch...
skip test_0_exception_in_cctor...
skip test_0_rethrow_stacktrace...
skip test_0_array_size...
skip test_0_ldflda_null_pointer...
Results: total tests: 68, all pass
Elapsed time: 0.027124 secs (0.027124, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/devirtualization.exe
Results: total tests: 6, all pass
Elapsed time: 0.000615 secs (0.000615, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/generics.exe
skip test_0_fullaot_linq...
skip test_0_synchronized_gshared...
skip test_0_fullaot_sflda_cctor...
skip test_0_fullaot_imt...
skip test_0_fullaot_variant_iface...
skip test_1_regress_constrained_iface_call_7571...
skip test_0_regress_constrained_iface_call_enum...
skip test_0_delegate_callvirt_fullaot...
skip test_0_constrained_partial_sharing...
Results: total tests: 63, all pass
Elapsed time: 0.057870 secs (0.057870, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/basic-simd.exe
Results: total tests: 183, all pass
Elapsed time: 0.010082 secs (0.010082, 0.000000)

Test run: image=/Users/bernhardu/work/mono/mono/mini/basic-vectors.exe
skip test_0_vector2_abs...
skip test_0_vector4_abs...
skip test_0_vector3_abs...
Results: total tests: 95, all pass
Elapsed time: 0.086282 secs (0.086282, 0.000000)

Overall results: tests: 867, 100% pass
```

### CODE REVIEW

**Comment by lewurm** [comment]
@monojenkins build interpreter

File: mono/mini/tramp-amd64.c [hunk 1]
```diff
@@ -982,79 +982,149 @@ gpointer
 mono_arch_get_enter_icall_trampoline (MonoTrampInfo **info)
 {
 #ifdef ENABLE_INTERPRETER
-	const int gregs_num = 6;
-	guint8 *start = NULL, *code, *exits [gregs_num], *leave_tramp;
+#ifdef TARGET_WIN32
```
Comment (joncham) [inline mono/mini/tramp-amd64.c]: There is `PARAM_REGS` in mini-amd64.h that seems to correspond to this

File: mono/mini/tramp-amd64.c [hunk 2]
```diff
@@ -982,79 +982,149 @@ gpointer
...
+	const int gregs_passed_via_regs = 6;
+#endif
+	const int gregs_num = 8;
+	const int fregs_num = 3;
+	guint8 *start = NULL, *code, *label_gexits [gregs_num], *label_fexits [fregs_num], *label_leave_tramp [3], *label_is_float_ret;
 	MonoJumpInfo *ji = NULL;
 	GSList *unwind_ops = NULL;
-	static int arg_regs[] = {AMD64_ARG_REG1, AMD64_ARG_REG2, AMD64_ARG_REG3, AMD64_ARG_REG4, AMD64_R8, AMD64_R9};
-	int i, offset = 0;
+	static int garg_regs[] = {AMD64_ARG_REG1, AMD64_ARG_REG2, AMD64_ARG_REG3, AMD64_ARG_REG4, AMD64_R8, AMD64_R9};
```
Comment (joncham) [inline mono/mini/tramp-amd64.c]: Similarly `param_regs`

File: mono/metadata/marshal.c [hunk 3]
```diff
@@ -3621,9 +3621,12 @@ mono_marshal_get_delegate_invoke_internal (MonoMethod *method, gboolean callvirt
 			mono_mb_emit_ldloc (mb, local_target);
 			for (i = 0; i < sig->param_count; ++i)
 				mono_mb_emit_ldarg (mb, i + 1);
-			mono_mb_emit_ldarg (mb, 0);
-			mono_mb_emit_ldflda (mb, MONO_STRUCT_OFFSET (MonoDelegate, extra_arg));
-			mono_mb_emit_byte (mb, CEE_LDIND_I);
+
```
Comment (vargaz) [inline mono/metadata/marshal.c]: Can't the interpreter handle this the same way the JIT does ?

File: mono/mini/tramp-amd64.c (hunk 1 above)
Comment (lewurm) [inline mono/mini/tramp-amd64.c]: thanks! :)

File: mono/metadata/marshal.c (hunk 3 above)
Comment (lewurm) [inline mono/metadata/marshal.c]: not exactly sure how the JIT handles it, but indeed, we can just remove the argument from the stack when we see a `MONO_CALLI_EXTRA_ARG`: https://github.com/mono/mono/pull/4430/commits/c66806a5ad54db6e973ffa649a36028ea153866c

thus, I removed the changes to `marshal.c` 👍

**Comment by lewurm** [comment]
@monojenkins build interpreter
